src/libraries/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -48,6 +49,31 @@ internal static bool CheckIriUnicodeRange(char highSurr, char lowSurr, out bool
             return false;
         }
 
+        internal static bool CheckIriUnicodeRange(uint value, bool isQuery)
+        {
+            if (value <= 0xFFFF)
+            {
+                return IsInInclusiveRange(value, '\u00A0', '\uD7FF')
+                    || IsInInclusiveRange(value, '\uF900', '\uFDCF')
+                    || IsInInclusiveRange(value, '\uFDF0', '\uFFEF')
+                    || (isQuery && IsInInclusiveRange(value, '\uE000', '\uF8FF'));
+            }
+            else
+            {
+                return ((value & 0xFFFF) < 0xFFFE)
+                    && ((value - 0xE0000) >= (0xE1000 - 0xE0000))
+                    && (isQuery || value < 0xF0000);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsInInclusiveRange(uint value, uint min, uint max)
+            => (value - min) <= (max - min);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsInInclusiveRange(char value, char min, char max)
+            => (uint)(value - min) <= (uint)(max - min);
```

Comment:
We should put these in a common class; there are places in HTTP we do similar things, and probably elsewhere in text processing too. CC @GrabYourPitchforks 

...  ideally JIT would just optimize this for us.